### ISSUE TITLE
Performance problem in "recommend by author" plugin

### ISSUE THREAD

**Issue by asmecher** [pkp/pkp-lib#5887]
The "Recommend By Author" plugin queries for an author across all journals in the installation. However, there is no index for that condition and large collections can get quite slow.

The issue was reported (by @ajnyga) in OJS 3.1.2; in OJS 3.2.0, the query appears to have been rewritten to be journal-constrained, which is probably not intentional.

**Comment by asmecher** [comment]
Also reported on the forum: https://forum.pkp.sfu.ca/t/ojs-3-1-2-plugin-of-recommended-articles-of-the-author-makes-the-platform-slower/60628/2

**Comment by mfelczak** [comment]
For a multi-tenant install or an install with unrelated journals, it would be preferable to keep this plugin to be constrained by journal or to add a setting to retain this functionality.

**Comment by ajnyga** [comment]
Was this fixed for 3.3?

**Comment by asmecher** [comment]
No, I don't think there's been any dev activity on this one yet.

**Comment by asmecher** [comment]
@mfelczak, this feature currently searches OJS's author records by given and last name, looking for exact matches. This performs badly but also doesn't align with our plans for author disambiguation using ORCIDs. Do the clients who are asking for this feature tend to use ORCIDs?

**Comment by mfelczak** [comment]
Hi @asmecher, I don't have a firm sense of ORCID adoption across our hosted journals as it's not something that we currently track. We could provide some stats though if that might be helpful, e.g. checking whether the ORCID plugin is enabled.

**Comment by jonasraoni** [comment]
I can confirm this plugin is a bit resource intensive. As I've a good place to test some improvements, I've assigned it to me.

**Comment by Ian-BD** [comment]
Has there much headway in improving the performance of this plugin? 

We run a large multi-journal site and was having some major performance issues which we've narrowed down to this.

For now we've disabled the plugin but if you have any suggestions as to how we can make some temporary workarounds while you work on improvements that'll be greatly appreciated.

**Comment by jonasraoni** [comment]
@Ian-BD I've just had time to update a similar plugin (https://github.com/pkp/pkp-lib/issues/8710). Last time I've skimmed the code, I saw some opportunities to optimize it, in case you get it fixed, feel free to contribute :)

**Comment by edersotto** [comment]
@jonasraoni you offered above that someone could contribute a fix, so here is what we found and a first PR.

Measured on a live journal (OJS 3.5.0-5, MariaDB 11.4) with **4,823 published articles, 25,877 authors, 401,286 rows in `author_settings`**:

| article | time | queries |
|---|---|---|
| 32 authors (worst found) | **11.9 s** | 1,441 |
| 21 authors | 6.4 s | 410 |
| average over 25 articles | 0.80 s | 183 |

Two separate causes, and I think they are worth treating separately.

### 1. The lookup has no index (pkp/pkp-lib#13174)

`filterByName()` searches `author_settings` **by value**, and the table is indexed only by `author_id` and the `(author_id, locale, setting_name)` unique key, so `EXPLAIN` reports `type=ALL` over 401,286 rows — twice per author of the article being read.

Adding the index that `publication_settings` already has, and **changing nothing else**, gives:

```
EXPLAIN   type=ALL rows=401286   ->   type=ref rows=443
32-author article   12.6 s -> 0.57 s
average of 25       0.80 s -> 0.07 s
```

PRs: pkp/pkp-lib#13174 and pkp/ojs#5743. The index took 0.9 s to create on that database.

### 2. The plugin still computes on every page view

Even indexed, the plugin runs a query per author, plus `Repo::publication()->get()` per result and a statistics aggregate, on **every view of the most visited page of the site**.

We rewrote it so the article page only reads: an author index (normalised name, plus ORCID) is kept in three tables, and a scheduled task refreshes a slice per run — never everything at once. That gives **16.6 ms** on a cache miss and **0.14 ms** on a hit, and rebuilding the whole journal takes 12 s.

Two things came out of it that are relevant here regardless of what you do with our code:

- **ORCID.** @asmecher wrote above that name matching "doesn't align with our plans for author disambiguation using ORCIDs". We match on ORCID **in addition** to the name, never instead of it, so enabling it can only add articles — for instance an author who has since changed surname.
- **Authors with no name at all.** The current code calls `filterByName('', '')` for them, which matches every other nameless author in the installation. On that journal there are **222 such authors across 97 published articles**, each getting ~3,285 irrelevant recommendations and ~100 s per page view. Skipping empty names is a small fix worth making even without the rest.

The rewrite is running in production on that journal and is at https://github.com/OJSBR/recommendByAuthor, with a regression suite (62 cases), unit tests and a Cypress spec. Happy to open it as a PR here against `stable-3_5_0` and `main` if you would like it — or to just contribute the empty-name fix on its own, if that is easier to land.
